zipline/__main__.py

```diff
@@ -17,6 +18,20 @@
 except NameError:
     __IPYTHON__ = False
 
+extension_args = {}
+
+
+def _parse_extension_arg(arg):
+    match = re.match(r'^(([^\d\W]\w*)(\.[^\d\W]\w*)*)=(.*)$', arg)
+    if match is None:
+        raise ValueError(
+            "invalid extension argument %s, must be in key=value form" % arg
+        )
+
+    name = match.group(1)
+    value = match.group(4)
+    extension_args[name] = value
```

Comment:
Rather than mutating a global value in place, I would probably write this function as:

```python
def parse_extension_args(args):
    out = {}
    for arg in args:
        name, value = parse_single_extension_arg(arg)
        out[name] = value
    return out
```

Then, in `main`, we can do:

```python
args = parse_extension_args(x)
set_global_extension_args(args)  # Put the args somewhere user-accessible.
```

The benefit of this approach is that its easier to test `parse_extension_args` in isolation, because it doesn't interact with any global state: it's just responsible for the business logic of converting a list of command line flags into a dictionary.

Reply:
As far as setting the actual global state goes, there are a bunch of options. The simplest option is to just set a dictionary (or namespace object) into a global variable somewhere public and documented like `zipline.extensions.extension_args` as @llllllllll suggested in https://github.com/quantopian/zipline/pull/2201/files#r192841333.

Slightly more complex would be make the exact location of the storage private (e.g., `zipilne.extensions._global_extension_args`, and expose a public function like `get_extension_args` that retrieves the global variable. For testing, it might be useful to provide a [contextmanager](https://docs.python.org/2/library/contextlib.html#contextlib.contextmanager) for contextually setting and re-setting the global state of the extension args.

Here's an example of what this might look like:

```python
# zipline.extension.extension.py
from contextlib import contextmanager


_extension_args = None


def get_extension_args():
    """Useful user-facing docstring.
    """
    if _extension_args is None:
        raise NotInAZiplineExtension("stahp")
    return _extension_args


def _set_extension_args(args):
    """Private, internal-facing docstring.
    """
    global extension_args
    # Return the old args so that they can be restored.
    old_args = extension_args
    extension_args = args
    return old_args


@contextmanager
def _extension_args_context(args):
    """Private, internal-facing docstring.
    """
    try:
        old = _set_extension_args(args)
        yield
    finally:
        _set_extension_args(old)
```

With this version, our `main()` would look something like:

```
def main(..., x):
    with _extension_args_context(parse_extension_args(x)):
        load_extensions(...)
    do_stuff(...)
```

and user code would interact with this system by doing:

```python
from zipline.extensions import get_extension_args

args = get_extension_args()
do_stuff_with(args)
```